app/src/main/java/org/projectbuendia/client/sync/controllers/IncrementalSyncPhaseRunnable.java

```diff
@@ -0,0 +1,182 @@
+
+        String syncToken = SyncAdapter.getLastSyncToken(providerClient, dbTable);
+
+        IncrementalSyncResponse<T> response;
+
+        do {
+            RequestFuture<IncrementalSyncResponse<T>> future = RequestFuture.newFuture();
+            createRequest(syncToken, future, future);
+            response = future.get();
+            ArrayList<ContentProviderOperation> ops =
+                    getUpdateOps(response.results, syncResult);
+            providerClient.applyBatch(ops);
+            LOG.i("Updated page of %s (%d db ops)", resourceType, ops.size());
+
+            // Update sync token
+            syncToken = response.syncToken;
+        } while (response.more);
+
+        SyncAdapter.storeSyncToken(providerClient, dbTable, response.syncToken);
+
+        afterSyncFinished(contentResolver, syncResult, providerClient);
+    }
+
+    // Mandatory callback
+
+    /**
+     * Produces a list of the operations needed to bring the local database in sync with the server.
+     */
+    protected abstract ArrayList<ContentProviderOperation> getUpdateOps(
+            T[] list, SyncResult syncResult);
+
+    // Optional callbacks
+
+    /** Called before any records have been synced from the server. */
+    protected void beforeSyncStarted(
+            ContentResolver contentResolver,
+            SyncResult syncResult,
+            ContentProviderClient providerClient) throws Throwable {}
+
+    /**
+     * Called after all records have been synced from the server, even if the number of synced
+     * records was zero.
+     */
+    protected void afterSyncFinished(
+            ContentResolver contentResolver,
+            SyncResult syncResult,
+            ContentProviderClient providerClient) throws Throwable {}
+
+    private void createRequest(
+            @Nullable String lastSyncToken,
+            Response.Listener<IncrementalSyncResponse<T>> successListener,
+            final Response.ErrorListener errorListener) {
+        OpenMrsConnectionDetails connectionDetails = App.getConnectionDetails();
+        Uri.Builder url = Uri.parse(connectionDetails.getBuendiaApiUrl()).buildUpon();
+        url.appendPath(resourceType);
+        if (lastSyncToken != null) {
+            url.appendQueryParameter("since", lastSyncToken);
+        }
+        GsonRequest<IncrementalSyncResponse<T>> request = new GsonRequest<>(
+                url.build().toString(),
+                new IncrementalSyncResponseType(clazz),
+                connectionDetails.addAuthHeader(new HashMap<String, String>()),
+                successListener,
+                wrapErrorListener(errorListener));
+        Serializers.registerTo(request.getGson());
+        request.setRetryPolicy(
+                new DefaultRetryPolicy(Common.REQUEST_TIMEOUT_MS_MEDIUM, 1, 1f));
+        connectionDetails.getVolley().addToRequestQueue(request);
+    }
+
+    private static class IncrementalSyncResponseType implements ParameterizedType {
```

Comment:
It's a static nested class because it doesn't know anything about the outer class it's contained in. See https://docs.oracle.com/javase/tutorial/java/javaOO/nested.html